File: day06.py
from typing import List

# Day 6: Wait For It

test_data = '''Time:      7  15   30
Distance:  9  40  200'''.split('\n')


def part1(data: List[str]):
    times = [int(t) for t in data[0].split()[1:]]
    distances = [int(d) for d in data[1].split()[1:]]
    total_ways = 1
    for t, d in zip(times, distances):
        total_ways *= sum([1 for i in range(1, t) if i * (t - i) > d])
    return total_ways


def part2(data: List[str]):
    times = [int(t) for t in data[0].replace(' ', '').split(':')[1:]]
    distances = [int(d) for d in data[1].replace(' ', '').split(':')[1:]]
    total_ways = 1
    for t, d in zip(times, distances):
        total_ways *= sum([1 for i in range(1, t) if i * (t - i) > d])
    return total_ways


def main():
    with open('inputs/day06.txt') as read_file:
        data = [x.rstrip('\n') for x in read_file.readlines()]

    part1_test_result = part1(test_data)
    assert part1_test_result == 288, f'Part 1 test input returned {part1_test_result}'
    part1_result = part1(data)
    assert part1_result == 1083852, f'Part 1 returned {part1_result}'

    part2_test_result = part2(test_data)
    assert part2_test_result == 71503, f'Part 2 test input returned {part2_test_result}'
    part2_result = part2(data)
    assert part2_result == 23501589, f'Part 2 returned {part2_result}'


if __name__ == '__main__':
    main()

File: rust-aoc/src/day06.rs
fn count_ways(times: Vec<usize>, distances: Vec<usize>) -> usize {
    let mut total_ways: usize = 1;
    for i in 0..times.len() {
        let t = times[i];
        let d = distances[i];
        let mut sum: usize = 0;
        for j in 0..t {
            if j * (t - j) > d {
                sum += 1;
            }
        }
        total_ways *= sum;
    }
    return total_ways;
}

fn split_to_ints(s: &str) -> Vec<usize> {
    return s.split_whitespace().collect::<Vec<&str>>().iter().skip(1).map(|el| el.parse().unwrap()).collect();
}

fn part1(data: &str) -> usize {
    let parts = data.split('\n').collect::<Vec<&str>>();
    let times = split_to_ints(parts[0]);
    let distances = split_to_ints(parts[1]);
    return count_ways(times, distances);
}

fn split_part2(s: &str) -> Vec<usize> {
    return s.replace(" ", "").as_str().split(":").collect::<Vec<&str>>().iter().skip(1).map(|el| el.parse().unwrap()).collect();
}


fn part2(data: &str) -> usize {
    let parts = data.split('\n').collect::<Vec<&str>>();
    let times: Vec<usize> = split_part2(parts[0]);
    let distances = split_part2(parts[1]);
    return count_ways(times, distances);
}

fn main() {
    let content = include_str!("../../inputs/day06.txt");
    let before = std::time::Instant::now();
    let result1 = part1(&content);
    assert_eq!(result1, 1083852);
    // println!("Result part 1: {result1}");
    let result2 = part2(&content);
    assert_eq!(result2, 23501589);
    // println!("Result part 2: {result2}");
    let elapsed = before.elapsed();
    println!("Day 6: {:?}", elapsed);
}
